java/ql/lib/semmle/code/java/frameworks/apache/Collections.qll

```diff
@@ -182,6 +204,695 @@ private class ApacheKeyValueModel extends SummaryModelCsv {
+          ".list;FixedSizeList;true;FixedSizeList;;;Element of Argument[0];Element of Argument[-1];value",
+          ".list;FixedSizeList;true;fixedSizeList;;;Element of Argument[0];Element of ReturnValue;value",
+          ".list;GrowthList;true;GrowthList;(List);;Element of Argument[0];Element of Argument[-1];value",
+          ".list;GrowthList;true;growthList;;;Element of Argument[0];Element of ReturnValue;value",
+          ".list;LazyList;true;LazyList;;;Element of Argument[0];Element of Argument[-1];value",
+          ".list;LazyList;true;lazyList;;;Element of Argument[0];Element of ReturnValue;value",
+          ".list;NodeCachingLinkedList;true;NodeCachingLinkedList;(Collection);;Element of Argument[0];Element of Argument[-1];value",
+          ".list;PredicatedList;true;PredicatedList;;;Element of Argument[0];Element of Argument[-1];value",
+          ".list;PredicatedList;true;predicatedList;;;Element of Argument[0];Element of ReturnValue;value",
+          ".list;SetUniqueList;true;SetUniqueList;;;Element of Argument[0];Element of Argument[-1];value",
+          ".list;SetUniqueList;true;SetUniqueList;;;Element of Argument[1];Element of Argument[-1];value",
+          ".list;SetUniqueList;true;setUniqueList;;;Element of Argument[0];Element of ReturnValue;value",
+          ".list;SetUniqueList;true;asSet;;;Element of Argument[-1];Element of ReturnValue;value",
+          ".list;SetUniqueList;true;createSetBasedOnList;;;Element of Argument[1];Element of ReturnValue;value",
+          ".list;TransformedList;true;TransformedList;;;Element of Argument[0];Element of Argument[-1];value",
+          ".list;TransformedList;true;transformingList;;;Element of Argument[0];Element of ReturnValue;value",
+          ".list;TransformedList;true;getList;;;Element of Argument[-1];Element of ReturnValue;value",
+          ".list;TreeList;true;TreeList;;;Element of Argument[0];Element of Argument[-1];value",
+          ".list;UnmodifiableList;true;UnmodifiableList;;;Element of Argument[0];Element of Argument[-1];value",
+          ".list;UnmodifiableList;true;unmodifiableList;;;Element of Argument[0];Element of ReturnValue;value"
+        ]
+  }
+}
+
+/**
+ * Value-propagating models for the package `org.apache.commons.collections4.map`.
+ */
+private class ApacheMapModel extends SummaryModelCsv {
+  override predicate row(string row) {
+    row =
+      ["org.apache.commons.collections4", "org.apache.commons.collections"] +
+        [
+          // Note that when lambdas are supported we should have more models for DefaultedMap, LazyMap, TransformedMap, TransformedSortedMap
+          ".map;AbstractHashedMap;true;AbstractHashedMap;(Map);;MapKey of Argument[0];MapKey of Argument[-1];value",
+          ".map;AbstractHashedMap;true;AbstractHashedMap;(Map);;MapValue of Argument[0];MapValue of Argument[-1];value",
+          ".map;AbstractHashedMap;true;convertKey;;;Argument[0];ReturnValue;value",
```

Comment:
```suggestion
          ".map;AbstractHashedMap;true;convertKey;;;Argument[0];ReturnValue;taint",
```

Reply:
I was thinking about the implementation in this class, which does return `Argument[0]` except in the case that it is null. I shouldn't have put `true` for the `subtypes` column. Presumably most overrides would implement a hash algorithm, which would be very hard to reverse, so I'm not sure it's worth modelling this as taint flow for subtypes.